servers/physics_2d/godot_joints_2d.cpp

```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
 	if (B && dynamic_B) {
 		B->apply_impulse(P, rB);
 	}
+	// from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L25
+	real_t i_sum_local = A->get_inv_inertia();
+	if (B) {
+		i_sum_local += B->get_inv_inertia();
+	}
+	i_sum = 1.0f/(i_sum_local);
+	if (enable_limits && B) {
+		real_t dist = B->get_transform().get_rotation() - A->get_transform().get_rotation();
+		real_t pdist = 0.0f;
+		if(dist > angular_limit_upper) {
+			pdist = angular_limit_upper - dist;
+		} else if(dist < angular_limit_lower) {
+			pdist = angular_limit_lower - dist;
+		}
+		real_t error_bias = powf(1.0f - 0.1f, 60.0f);
+		// calculate bias velocity
+		bias_velocity = CLAMP((-1.0f - powf(error_bias, p_step))*pdist/p_step, -get_max_bias(), get_max_bias());
+
+		// If the bias velocity is 0, the joint is not at a limit.
+		if(bias_velocity >= -CMP_EPSILON || bias_velocity <= CMP_EPSILON) {
+			is_joint_at_limit = false;
+		} else {
+			is_joint_at_limit = true;
+		}
+	} else {
+		bias_velocity = 0;
+	}
 
 	return true;
 }
 
 void GodotPinJoint2D::solve(real_t p_step) {
 	// compute relative velocity
 	Vector2 vA = A->get_linear_velocity() - custom_cross(rA - A->get_center_of_mass(), A->get_angular_velocity());
-
+	
 	Vector2 rel_vel;
 	if (B) {
 		rel_vel = B->get_linear_velocity() - custom_cross(rB - B->get_center_of_mass(), B->get_angular_velocity()) - vA;
 	} else {
 		rel_vel = -vA;
 	}
 
+	// from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c#L45
+	// and https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L61
```

Comment:
```suggestion
	// Taken from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c#L45
	// and https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L61
```
Same here for permalink